### PR TITLE
Pandoc and source code injection and linting! (Oh my!)

### PR DESCRIPTION
**Pull request by ngscheurich**
So, this is kind of a biggie so I'll try and organize my thoughts.

# Pandoc

Since we're searching out a new build tool (#28), I thought I'd experiment some with [Pandoc](https://pandoc.org/). For those unfamiliar with Pandoc, it bills itself thusly:

> If you need to convert files from one markup format into another, pandoc is your swiss-army knife. Pandoc can convert documents in (several dialects of) Markdown, reStructuredText, textile, HTML, DocBook, LaTeX, MediaWiki markup, TWiki markup, TikiWiki markup, Creole 1.0, Vimwiki markup, OPML, Emacs Org-Mode, Emacs Muse, txt2tags, Microsoft Word docx, LibreOffice ODT, EPUB, or Haddock markup to…

The description then goes on to list a bunch of output formats including EPUB, HTML, and PDF. The Pandoc site has a documentation section dedicated to creating ebooks, and uses the (relatively) well-known [Pro Git 1st Edition](https://git-scm.com/book/en/v1) as an example—this gave me high confidence that Pandoc would be a viable candidate.

## An alternative project layout

To make things easier to process with Pandoc, I thought it prudent to reorganize the project—specifically the `book` directory—a bit. The new tree looks something like:

```
book
├── 1_introduction
│   ├── i_what_is_a_roguelike.m
│   ├── ii_what_is_nethack.md
│   ├── …
├── 2_generating_a_character
│   ├── 1_the_title_screen.md
│   ├── …
├── 3_dungeon_generation
│   └── 8_generating_a_room.md
└── title.txt
```

In addition to making Pandoc processing easier, I feel like we also get an increase in ease of cognitive and programmatic sorting of the source files. I also renamed the `images` directory `figures` to more closely align with the language in the Pandoc documentation.

## Image (figure) markup

Since Pandoc can natively embed images marked up in standard Markdown, I replaced all the lines in the first two sections that looked like this:

> `\includegraphics[width=\linewidth]{images/nethack.png}`

With lines that look like this:

> ``

*Note: We should probably add some alt text (which will be converted to figure captions) to each image.*

## Linting

During the process of making the aforementioned changes, I started to consider that we'd benefit from having a linting tool set up so that all of our sample code retains a consistent style. Following that train of thought a bit, I also thought it would be nice to be able to edit example code with correct syntax highlighting and all of the other niceties a well-configured text editor affords. So, we come now to…

# Source code injection

Editing each bit of sample code in its own file seemed like something that would fit the bill and the Pandoc-powered build tool I put together seemed like a good place to hook something up to handle this. After wiring everything up, I ended up with a `code` directory to hold all the sample code and was able to change stuff like this:

> ````ruby
> require "curses" # require the curses gem
> include Curses   # mixin curses
>
> # The next three methods are provided by including the Curses module.
>
> init_screen      # starts curses visual mode
> getch            # reads a single character from stdin
> close_screen     # closes the ncurses screen
> ```

To stuff like this:

> `!{code/2_1_title_screen_1.rb}`

I didn't swap out the code in all the files but did enough to hopefully provide a solid example of the general workflow.

Cool. Now sample code is super easy to lint, which brings us to…

# RuboCop

RuboCop is the de facto linter for maintaining Ruby code style, so I figured that would be the way to go for this project. In order to avoid bikeshedding (including of the personal variety), I simply copied down [thoughtbot's RuboCop config](https://github.com/thoughtbot/guides/blob/master/style/ruby/.rubocop.yml), updated it for the most current version of RuboCop, and fixed a small typo. There were somewhere in the neighborhood of 14 violations across the sample code and the example app, so I fixed those and called it a day.

# All together now

After merging this code, you should be able to run `exe/compile` and wind up with a compiled EPUB in the `release` directory. More output formats (and a title page) can easily be added moving forward.

# Conclusion

I understand that this is probably way too much change to consider for a single pull request, but that's just sort of how everything flowed from my head-cold-addled mind into my fingertips. I hope that if we decide not to merge these changes, this code will at least spawn some interesting discussion and help inform other changes to the project.

If you made it this far, thanks for reading!

### CODE REVIEW

**Comment by ngscheurich** [comment]
Quick note: You can generate an EPUB version of the book by running `exe/compile`. Currently there is no title page, but that can be easily rectified moving forward.

**Comment by richdouglasevans** [comment]
Oh my! :open_mouth: 

:smile: 

I heartily approve of this PR... and the actual text of the PR itself: lovely.

/me rushes away to checkout this branch and "pandoc all the shizzes"

**Comment by ngscheurich** [comment]
Thanks for the vote of confidence, @richdouglasevans! Let me know if you have any issues or suggestions.

**Comment by ngscheurich** [comment]
For those interested, should probably add: You'll need the [`pandoc-ruby` gem](https://github.com/alphabetum/pandoc-ruby) installed to run the `compile` script.

File: book/sample.md [hunk 1]
```diff
@@ -1,18 +0,0 @@
-% Write Yourself a Roguelike
```
Comment (tvanderpol) [inline book/sample.md]: Just to clarify and make the thinking explicit - there's no need for these sample files because the whole book is available now, correct?

File: example_app/attribute_generator.rb [hunk 2]
```diff
@@ -32,7 +32,7 @@ def increment_random_attribute
   def next_random_attribute
     x = rand(100)
 
-    base_attributes.keys.find do |key|
+    base_attributes.keys.detect do |key|
```
Comment (tvanderpol) [inline example_app/attribute_generator.rb]: Unless this is an error in the script, it might be best to separate code changes from this already substantial structural change? If nothing else, it'll make for a cleaner history with regards to what PR made which change.

And out of personal curiosity - why change .find to .detect?

File: example_app/dungeon.rb [hunk 3]
```diff
@@ -101,12 +101,12 @@ def bottom
 class DungeonGenerator
   DEFAULT_WIDTH = 80
   DEFAULT_HEIGHT = 21
-  DEFAULT_TILESET_NAME = "default"
+  DEFAULT_TILESET_NAME = "default".freeze
```
Comment (tvanderpol) [inline example_app/dungeon.rb]: As before (and there's a couple more changes in this file and `example_app/game.rb` as well) - are these changes in scope for this PR or should they be separated?

File: book/sample.md (hunk 1 above)
Comment (ngscheurich) [reply book/sample.md]: Yep, that's my thinking. I'm unsure of the original utility of the sample files, but I assume they were to offer potential customers a glimpse of the book. Since now people can read the entire book for free, I say trash the samples.

File: example_app/attribute_generator.rb (hunk 2 above)
Comment (ngscheurich) [reply example_app/attribute_generator.rb]: I agree—moving the code changes to a separate PR makes the intent of both requests much clearer. What method would you propose we use to move the code changes out of this PR?

Regarding `find` vs. `detect`: That was to satisfy the linter, which is currently using Thoughtbot's config. I don't have a preference either way, so if you like `find`, let's use it and update the config.

File: example_app/dungeon.rb (hunk 3 above)
Comment (ngscheurich) [reply example_app/dungeon.rb]: See above; agreed on separating out code style changes.

**Comment by ngscheurich** [comment]
Not sure what the merge conflicts are about… The branch merges fine locally. We may need to do the rebase/merge manually.

**Comment by ngscheurich** [comment]
As per the wise suggestion made by @tvanderpol to split this PR into separate ones, I've created #33 and #34. Closing this PR in favor of the aforementioned.
